### DISCUSSION TITLE
Use of parcels.FieldSet.from_croco()

### DISCUSSION THREAD

**Discussion by erikvansebille** [Parcels-code/Parcels#1752] [Q&A]
### Question

Parcels v3.1.0 added a `parcels.FieldSet.from_croco()` method, with the goal to support native, under-the-hood advection in CROCO sigma grids. See [here](https://docs.oceanparcels.org/en/latest/examples/tutorial_croco_3D.html) for the tutorial and specifically [here for 

However, it's unclear whether [the algorithm used](https://docs.oceanparcels.org/en/latest/examples/tutorial_croco_3D.html#The-algorithms-used) is actually the correct one for Parcels. The CROCO team has recently explained us that "it is not possible to switch exactly from depths to sigma coordinates and therefore to go back and forth between sigma and Z." I hope they can provide some more information below, because I think it's good if every user knows about this issue, and we can hopefully come up with a working solution together.

**Comment by JamiePringle** [comment]
You may find this thread from the ROMS mailing list to be helpful. CROCO is a branch of ROMS. https://www.myroms.org/forum/viewtopic.php?t=6639 . I am puzzled by the CROCO teams response, unless they are talking about issues caused by the split 2D/3D time step.  I would be very interested in what they say. 

Jamie

**Comment by stefraynaud** [comment]
I think @Mesharou is the most appropriate person to address the issue of Lagrangian simulations in CROCO.

@JamiePringle, I can confirm that there is no difficulty in converting “s” coordinates to “z”.

One question is whether to use the “omega” variable or the “w” variable for vertical velocities.

@Mesharou used omega in [Pyticles](https://github.com/Mesharou/Pyticles). omega is output at point W. Lagrangian calculations are performed entirely in “s” coordinates, and we only switch back to “z” to export the results.

It's probably easier to use the “w” variable, which is output at the RHO point. I'm not enough an expert of Parcels to know which approach best respects the existing 3D advection kernels.

**Comment by erikvansebille** [comment]
Thanks for the response, @stefraynaud. I think what to me was the most striking part of your earlier email was the point that the vertical coordinate is not linear in `sigma`: it is of type `ocean_s_coordinate_g1`. See also the [CROCO documentation](https://croco-ocean.gitlabpages.inria.fr/croco_doc/model/model.grid.html#vertical-grid-parameters), and the [CF conventions for automatic conversion](
http://cfconventions.org/cf-conventions/cf-conventions.html#_ocean_s_coordinate_generic_form_1)

The transition from `sigma` to `Z` therefore requires variables and parameters other than only `H`. This is something I wasn't aware of before.

However, I don't think it's impossible to put this conversion in Parcels too. It just requires users to also specify a `depth_c` constant, a 1-dimensional `C` Field, and a three-dimensional (horizontal-time) `eta` Field. When we have all these, we should be able to compute the conversion under the hood? 

See also https://github.com/OceanParcels/Parcels/issues/1763, in which I propose to implement this conversion.

**Reply by erikvansebille** [reply]
I've started to play around with the nonlinear scaling in `ocean_s_coordinate_g1`. The difference between the full `sigma-to-z` conversion and the simple `z=sigma*H` assumption can indeed be quite substantial, see first panel of the plot below
![image](https://github.com/user-attachments/assets/802b3a42-789e-44f1-9316-835ad1863818)

There's also a substantial difference between using the `Cs_w` values from the output file, and calculating $C_s(\sigma)$ using the equation ("Old transformation") at https://croco-ocean.gitlabpages.inria.fr/croco_doc/model/model.grid.html#vertical-grid-parameters

However, I have trouble inverting this equation, because of the quite complicated structure of $C_s(\sigma)$. Even when assuming $\theta_b=0$ and $\theta_s=1$ (as in the example file I use), this is nontrivial. 

Does anyone know how the inverse operation (so calculating `sigma` when `z` is known) is done in CROCO?

**Reply by stefraynaud** [reply]
I suspect that it's not invertible and if so, you must interpolate I guess.

**Reply by JamiePringle** [reply]
@erikvansebille @stefraynaud I will look into this in the next week or so, using the ROMS generalized vertical coordinates -- has crocco rolled its own separate transforms?

Jamie

**Reply by erikvansebille** [reply]
> I suspect that it's not invertible and if so, you must interpolate I guess.

Thanks for this suggestion to use interpolation, @stefraynaud. That indeed works very well. See below a figure where I compare the linearly interpolated $C_s(\sigma)$ (orange) to the explicitly calculated $C_s(\sigma)$ (blue). Hardly a difference to be seen. 

However, this doesn't entire solve the inversion problem. Because we would still need an equation for $C_s(z)$. Because the inversion we would(?) need to implement is first
$$z_0(x, y, z) = \frac{z-\zeta(x, y, t)}{1+\frac{\zeta(x, y, t)}{h(x, y)}}$$
and then
$$\sigma(x, y, z, t) = \frac{z_0(x, y, z) - (h(x, y) - h_c)C_s(z)}{h_c}$$


![image](https://github.com/user-attachments/assets/905d52c3-15d0-4ca6-aaea-03695438e7f4)
